lemming.go

```diff
@@ -55,8 +57,56 @@ type Device struct {
 	stopped chan struct{}
 }
 
+// DevOpt is an interface that is implemented by options that can be handed to New()
+// for the device.
+type DevOpt interface {
+	isDevOpt()
+}
+
+// deviceConfig is a wrapper for an input OpenConfig RFC7951-marshalled JSON
+// configuration for the device.
+type deviceConfig struct {
+	// json is the contents of the JSON document (prior to unmarshal).
+	json []byte
+}
+
+// isDevOpt marks deviceConfig as a device option.
+func (*deviceConfig) isDevOpt() {}
+
+// DeviceConfig sets the startup config of the device to c.
+// Today we do not allow the configuration to be changed in flight, but this
+// can be implemented in the future.
+//
+// This DeviceOption is intended for standalone device testing.
+func DeviceConfig(c []byte) *deviceConfig {
+	return &deviceConfig{json: c}
+}
+
+// tlsCreds returns TLS credentials that can be used for a device.
+type tlsCreds struct {
+	c credentials.TransportCredentials
+}
+
+// TLSCredsFromFile loads the credentials from the specified cert and key file
+// and returns them such that they can be used for the gNMI and gRIBI servers.
+func TLSCredsFromFile(certFile, keyFile string) (*tlsCreds, error) {
+	t, err := credentials.NewServerTLSFromFile(certFile, keyFile)
+	if err != nil {
+		return nil, err
+	}
+	return &tlsCreds{c: t}, nil
+}
+
+// TLSCreds returns a wrapper of TransportCredentials into a DevOpt.
+func TLSCreds(c credentials.TransportCredentials) *tlsCreds {
+	return &tlsCreds{c: c}
+}
+
+// IsDevOpt implements the DevOpt interface for tlsCreds.
+func (*tlsCreds) isDevOpt() {}
+
 // New returns a new initialized device.
-func New(lis net.Listener, targetName, zapiURL string, opts ...grpc.ServerOption) (*Device, error) {
+func New(lis net.Listener, targetName, zapiURL string, opts ...DevOpt) (*Device, error) {
```

Comment:
i prefer how ygnmi does options, the code is much simpler. 

https://github.com/openconfig/ygnmi/blob/main/ygnmi/ygnmi.go#L171

Reply:
Done.